YouTube Regrets Extention page

mmmavis opened the pull request:
Closes #5081 

https://foundation-s-issue-5081-x9u2kj.herokuapp.com/campaigns/regrets-reporter

### To code reviewer

I basically created a new page type `YoutubeRegretsReporterPage` and make it available as child page of `CampaignIndexPage`. Stakeholders are fine with the content being hardcoded as the page is unlikely to change after launch.

There was some code refactor involved:
- created a base template for YouTube Regrets page and this RegretsReporter page
- extracted some SCSS into its own file
- (see inline comments)

---

### To design reviewer

**Noise texture**
Mockup has the noise background texture only covering the upper part of the page. However, in order to build it like the mockup we need to place a background image that's at least 4000px tall (to account for the scroll distance) and has a gradient with noise texture and gradually fades out at the bottom. This image will probably end up being multiple MBs big so it's better not to include it for performance sake. I came up with a way to add the noise texture using a small SVG but I can only make it cover the entire page (instead of just the top of the page). If this looks too distracting, we can remove the texture so the page background is just a black to white gradient. Just let me know!

**Newsletter signup form**
The form doesn't look 100% like the mockup. Reason: I used our existing signup component in the codebase to build it. Altering that component just for this page isn't really worth effort so 😁

mmmavis commented on network-api/networkapi/wagtailpages/templates/wagtailpages/pages/youtube_regrets_base.html:
```diff
@@ -0,0 +1,57 @@
+{% extends "pages/base.html" %}
```
base template for YouTube Regrets page and the RegretsReporter page

mmmavis commented on network-api/networkapi/wagtailpages/templates/wagtailpages/pages/youtube_regrets_page.html:
```diff
@@ -1,4 +1,4 @@
-{% extends "pages/base.html" %}
```
refactored

mmmavis commented on source/js/foundation/pages/youtube-regrets/index.js:
```diff
@@ -11,22 +11,29 @@ const BLOCK_SPEED_FACTOR = 0.8;
 const RING_SPEED_FACTOR = 0.6;
 
 let elements = {
-  introViewport: `#view-youtube-regrets .intro-viewport`,
```
updated selector names

mmmavis commented on source/sass/views/youtube-regrets-intro.scss:
```diff
@@ -0,0 +1,221 @@
+// YouTube Regrets tunnel intro
```
same code but refactored into its own file

mmmavis commented on source/sass/views/youtube-regrets.scss:
```diff
@@ -12,231 +22,6 @@
     );
   }
 
-  #nav-newsletter-form-wrapper.faded-in {
```
refactored

mmmavis commented on tasks.py:
```diff
@@ -25,6 +25,7 @@
         "--ignore=network-api/networkapi/wagtail_l10n_customization/*",
         "--ignore=network-api/networkapi/settings.py",
         "--ignore=network-api/networkapi/wagtailpages/__init__.py",
+        "--ignore=network-api/networkapi/wagtailpages/templates/wagtailpages/pages/youtube_regrets_reporter_page.html",
```
This line is from Theo. See: https://github.com/mozilla/foundation.mozilla.org/issues/5081#issuecomment-685759249

mmmavis commented on source/sass/views/youtube-regrets-reporter.scss:
```diff
@@ -0,0 +1,314 @@
+#view-youtube-regrets-reporter {
```
there's just a ton of overrides to deal with so the code here is not the cleanest 🙈

mmmavis commented on source/js/foundation/pages/youtube-regrets/index.js:
```diff
@@ -11,22 +11,29 @@ const BLOCK_SPEED_FACTOR = 0.8;
 const RING_SPEED_FACTOR = 0.6;
 
 let elements = {
```
I wanted to rename this file to `tunnel.js` but wanted to do that after the code review as renaming a file can make code review harder.

kristinashu commented:
Wow Mavis it's looking really good! I've listed a couple type issues below where I think the comps look better and more legible than the PR. 

Maybe we can chat about the background? I like your creative solution but the texture does feel a little busy behind all the text on the page. If you did just do the black to white gradient then could the background icons be added in?

(screen cap of icon from comp)
<img width="187" alt="image" src="https://user-images.githubusercontent.com/1068783/92553212-3e343580-f217-11ea-9ebd-02127fce4dc8.png">

Also I've only seen [this desktop comp](https://lindsay970727.invisionapp.com/console/share/T41BYHZU8Y/479637009), are there tablet and mobile version I should reference?

Type issues:

Can you please center align these? Having it left aligned when the numbers are center aligned makes them feels a little detached.

![01](https://user-images.githubusercontent.com/1068783/92552914-97e83000-f216-11ea-8784-ec0386a5c327.jpg)

Can you please increase the size and line height of the body text? I feels a little small in the context of this page.

![02](https://user-images.githubusercontent.com/1068783/92552922-9a4a8a00-f216-11ea-9ee5-e2353535caa8.jpg)

mmmavis commented:
@kristinashu thanks for the feedback. I've updated everything you mentioned. Agreed that the noise texture made text harder to read. It's been removed now.

As for mockup reference - please check the Sketch file that I dropped in in the GDrive. (Sorry I already lost the link lol but it's the folder you linked me to before)

Pomax commented on network-api/networkapi/wagtailpages/templates/wagtailpages/pages/youtube_regrets_base.html:
```diff
@@ -0,0 +1,57 @@
+      <div class="ring"></div>
+      <div class="ring"></div>
+      <div class="ring"></div>
+      <div class="ring"></div>
```
☎️ ♬♩♫

kristinashu commented:
Ah wonderful Mavis! I definitely prefer the updated background, thank you!

I think you did all the text updates except for increasing the body copy here (screen cap from PR):
![image](https://user-images.githubusercontent.com/1068783/92633590-d833c680-f287-11ea-99bb-60a7e3fa8ae5.png)

Could you please make it match your other updated body copy (screen cap from PR)?
![image](https://user-images.githubusercontent.com/1068783/92633661-f26da480-f287-11ea-92a3-e1dfa47fa883.png)

Pomax commented on source/js/foundation/pages/youtube-regrets/regrets-reporter-timeline.js:
```diff
@@ -0,0 +1,91 @@
+      let windowWidth =
+        window.innerWidth || document.documentElement.clientWidth;
+      let lastItemShown = this.checkIfInView(
+        document.querySelector(".timeline .segment:last-of-type")
```
probably worth caching this element in a `const` outside the event listener? Also, we already have `timeline`, so we can query that directly:
```
const last = timeline.querySelector(`.segment:last-of-type`);
```

mmmavis replied:
👍

Pomax commented on source/js/foundation/pages/youtube-regrets/regrets-reporter-timeline.js:
```diff
@@ -0,0 +1,91 @@
+import utility from "../../../utility.js";
+
+let elements = {
+  timeline: `#view-youtube-regrets-reporter .timeline-wrapper .timeline`,
+  segments: `#view-youtube-regrets-reporter .timeline-wrapper .timeline .segment`,
+  button: `#view-youtube-regrets-reporter .timeline-wrapper .btn-timeline-next`,
+};
```
It feels like this might be overkill for a piece of UI with only three "parts" to it, but if we're basically going to "build and forget" then this is fine. If we need to work on this in the future, we'll have to let people know that this _will_ require more time due to the fact that it will require rewriting to a cleaner form before we do any more work.

mmmavis replied:
@Pomax did you mean this doesn't need to be its own module? Or that we don't need to hold `timeline`, `segments`, and `button` in a `elements` object?

Pomax replied:
the latter - it doesn't offer as much benefit as it did for the youtube regret stories page, which had like 10 elements or something

Pomax commented on source/js/foundation/pages/youtube-regrets/regrets-reporter-timeline.js:
```diff
@@ -0,0 +1,91 @@
+
+        if (windowWidth + offset >= this.timeline.offsetWidth) {
+          // show the last item
+          offset = this.timeline.offsetWidth - windowWidth;
```
this probably also needs to take the button's width into account, so there's no overlap:

![image](https://user-images.githubusercontent.com/177243/92638002-7d519d80-f28e-11ea-9eae-78a1d132b31c.png)

mmmavis replied:
nice catch!

Pomax commented on network-api/networkapi/wagtailpages/templates/wagtailpages/pages/youtube_regrets_reporter_page.html:
```diff
@@ -0,0 +1,245 @@
+          <img src="{% static "_images/youtube-regrets/regrets-reporter/how-it-works-step2.png" %}" />
+          <p class="d-flex flex-column align-items-center">
+            <span class="step-number my-3">2</span>
+            {% if request.LANGUAGE_CODE == "en" %}
+              Report the video and recommendations that led you to it
+            {% else %}
+              {% trans "Report the video* and recommendations that led you to it" %}
+            {% endif %}
```
This page is not currently getting localized, due to https://github.com/mozilla/foundation.mozilla.org/pull/5121/files#r482310695, so the string is going to be the same for every user. If not, and that `*` is supposed to be a footnote, that should most definitely be a real link, to the part of the page that explains that footnote. Or even better: not be a footnote at all, since this is a web page and not printed media =)

```suggestion
            {% trans "Report the video and recommendations that led you to it" %}
```

mmmavis replied:
updated!

Pomax commented on network-api/networkapi/wagtailpages/templates/wagtailpages/pages/youtube_regrets_reporter_page.html:
```diff
@@ -0,0 +1,245 @@
+          </p>
+        </div>
+      </div>
+      {% if request.LANGUAGE_CODE != "en" %}
+        <div class="row py-md-4">{% trans "*Please note that only English-language videos will be included in our analysis." %}
+        </div>
+      {% endif %}
```
if possible, we should put this in the same block where we ask for videos.

mmmavis replied:
moved!

Pomax reviewed (changes requested):
some notes and questions

mmmavis commented on source/js/components/join/join.jsx:
```diff
@@ -34,7 +34,9 @@ export default class JoinUs extends React.Component {
       lang: getCurrentLanguage(),
       hideLocaleFields:
-        props.formPosition === `header` || props.formPosition === `footer`,
+        props.formPosition === `header` ||
+        props.formPosition === `footer` ||
+        props.formPosition === `youtube-regrets-reporter`,
```
totally agreed!

mmmavis commented on source/js/foundation/pages/youtube-regrets/index.js:
```diff
@@ -233,17 +240,19 @@ class YouTubeRegretsTunnel {
-      navNewsletter.buttonDesktopClickHandler(event)
-    );
+    if (this.includeNewsletterButtons) {
+      elements.newsletterButtonDesktop[0].addEventListener(`click`, (event) =>
```
I feel the same. Will file a new ticket for that to make sure the code changes don't break other modules that also make use of the `utility.checkAndBindDomNodes` function

mmmavis replied:
Filed: https://github.com/mozilla/foundation.mozilla.org/issues/5199

mmmavis commented on source/js/foundation/pages/youtube-regrets/index.js:
```diff
@@ -233,17 +240,19 @@ class YouTubeRegretsTunnel {
-      }
-    });
+      elements.newsletterButtonMobile[0].addEventListener(`click`, (event) => {
+        if (navNewsletter.getShownState()) {
```
Thanks!

mmmavis commented:
@Pomax @kristinashu this is ready for another review!

kristinashu reviewed (approved):
Sooo good!

Pomax commented on network-api/networkapi/wagtailpages/templates/wagtailpages/pages/youtube_regrets_reporter_page.html:
```diff
@@ -0,0 +1,241 @@
+            <span class="step-number my-3">2</span>
+            {% trans "Report the video and recommendations that led you to it" %}
+          </p>
+          {% if request.LANGUAGE_CODE == "en" %}
```
```suggestion
          {% if request.LANGUAGE_CODE != "en" %}
```

mmmavis replied:
Ooops!

Pomax reviewed (changes requested):
one thing~

mmmavis commented:
🎉 Thanks, team!
